### PR TITLE
Some work on date/time and network converters

### PR DESCRIPTION
**Pull request by roji**

### CODE REVIEW

File: src/Npgsql/Internal/Converters/Networking/MacaddrConverter.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+sealed class MacaddrConverter : PgBufferedConverter<PhysicalAddress>
+{
+    public override bool CanConvert(DataFormat format, out BufferingRequirement bufferingRequirement)
+    {
+        bufferingRequirement = BufferingRequirement.FixedSize;
+        return base.CanConvert(format, out _);
+    }
+
+    public override Size GetSize(SizeContext context, PhysicalAddress value, ref object? writeState)
+        => value is null // TODO: Remove this
```
Comment (roji) [inline src/Npgsql/Internal/Converters/Networking/MacaddrConverter.cs]: Note that null values end up here - ToBufferRequirements passes `null!` which gets converted to `default!` in GetSizeAsObject (so this seems like a bug with buffered converters over reference types).

File: src/Npgsql/Internal/AdoTypeInfoResolver.cs [hunk 2]
```diff
@@ -195,6 +211,19 @@ static void AddInfos(TypeInfoMappingCollection mappings)
                 static (options, mapping, _) => mapping.CreateInfo(options, new UInt32Converter()), isDefault: true);
         }
 
+        // TODO: Move out to opt-in resolver
+
+        // macaddr
+        mappings.AddType<PhysicalAddress>(DataTypeNames.MacAddr,
+            static (options, mapping, _) => mapping.CreateInfo(options, new MacaddrConverter()), isDefault: true);
+        mappings.AddType<PhysicalAddress>(DataTypeNames.MacAddr8,
```
Comment (NinoFloris) [inline src/Npgsql/Internal/AdoTypeInfoResolver.cs]: Probably PhysicalAddress <-> macaddr8 should be default as well. To make GetValue() work

For parameters, when we have duplicate registrations like this, the first one will win wrt resolving based on just a clr type.

File: src/Npgsql/Internal/Converters/Networking/MacaddrConverter.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+        var bytes = new byte[len];
+        reader.CopyTo(bytes);
+        return new PhysicalAddress(bytes);
+    }
+
+    protected override void WriteCore(PgWriter writer, PhysicalAddress value)
+    {
+        var bytes = value.GetAddressBytes();
+        // TODO: Should we have an optimized overload that just accepts ReadOnlySpan?
```
Comment (NinoFloris) [inline src/Npgsql/Internal/Converters/Networking/MacaddrConverter.cs]: Yep absolutely!

File: src/Npgsql/Internal/Converters/Networking/MacaddrConverter.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    public override Size GetSize(SizeContext context, PhysicalAddress value, ref object? writeState)
+        => value is null // TODO: Remove this
+            ? Size.Zero
+            : value.GetAddressBytes().Length;
+
+    protected override PhysicalAddress ReadCore(PgReader reader)
+    {
+        var len = reader.Current.Size.Value;
```
Comment (NinoFloris) [inline src/Npgsql/Internal/Converters/Networking/MacaddrConverter.cs:13]: Let me just add back the CurrentSize property for this. Yuck

File: src/Npgsql/Internal/Converters/Networking/MacaddrConverter.cs (hunk 1 above)
Comment (NinoFloris) [reply src/Npgsql/Internal/Converters/Networking/MacaddrConverter.cs]: As discussed offline, fixed size conversions shouldn't have to inspect the value. Just remove the CanConvert override to make it do the expected thing.

`PgBufferedConverter` defaults to `BufferingRequirement.Value`
`PgStreamingConverter` defaults to `BufferingRequirement.None`

File: src/Npgsql/Internal/Converters/Networking/InetConverter.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        bufferingRequirement = BufferingRequirement.FixedSize;
+        return base.CanConvert(format, out _);
+    }
+
+    public override Size GetSize(SizeContext context, IPAddress value, ref object? writeState)
+        => value is null // TODO: Remove this
+            ? Size.Zero
+            : value.AddressFamily switch
+            {
+                AddressFamily.InterNetwork => 8,
```
Comment (NinoFloris) [inline src/Npgsql/Internal/Converters/Networking/InetConverter.cs]: Same problem as discussed.

File: src/Npgsql/Internal/Converters/Networking/MacaddrConverter.cs (hunk 3 above)
Comment (roji) [reply src/Npgsql/Internal/Converters/Networking/MacaddrConverter.cs]: Done

**Review by NinoFloris — APPROVED** [review]
LGTM, depending on whether you still want to add NpgsqlInet you might want to exclude PgInet

**Comment by roji** [comment]
@NinoFloris rebased and redid the network types as per #5136 (I tried not to introduce creep but implementing the existing setup started to get really annoying, and as we want to change anyway...).

In any case, take a look at the last commit; two tests still fail, you may want to take a look at some point. After this I'll go on to work on other various converters.

File: src/Npgsql/Internal/Converters/Networking/MacaddrConverter.cs (hunk 4 above)
Comment (NinoFloris) [reply src/Npgsql/Internal/Converters/Networking/MacaddrConverter.cs:13]: It should exist now!

File: src/Npgsql/Internal/Converters/Temporal/DateTimeConverterResolvers.cs [hunk 6]
```diff
@@ -32,19 +32,17 @@ public override PgConverterResolution Get(DateTime value, PgTypeId? expectedPgTy
...
-
-            // We coalesce with expectedPgTypeId to throw on unknown type ids.
-            return GetDefault(expectedPgTypeId ?? _timestampTz);
+                    "Note that it's not possible to mix DateTimes with different Kinds in an array/range.", nameof(value))
+                : GetDefault(expectedPgTypeId ?? _timestampTz);
         }
 
-        if (expectedPgTypeId == _timestampTz)
+        if (expectedPgTypeId == _timestampTz
+            && !(_dateTimeInfinityConversions && (value == DateTime.MinValue || value == DateTime.MaxValue)))
```
Comment (NinoFloris) [inline src/Npgsql/Internal/Converters/Temporal/DateTimeConverterResolvers.cs:46]: Probably want this check on the values with its own exception at the start? (it's missing on the utc side atm)

Comment (roji) [reply src/Npgsql/Internal/Converters/Temporal/DateTimeConverterResolvers.cs:46]: I think DateTime.{Min,Max}Value are always Unspecified no?

In any case, I'd go ahead and merge, we can fix later (and hopefully if something's missing we should see it through test failures).

Comment (NinoFloris) [reply src/Npgsql/Internal/Converters/Temporal/DateTimeConverterResolvers.cs:46]: Those specific values are, yes, however doing `DateTime.MaxValue.ToUniversalTime()` should still IMO work. Just as we convert for min/max instants in nodatime.

Resultingly I would say the inverse should throw in both cases as well.

**Pull request merged**